lib/improver/blending/spatial_weights.py

```diff
@@ -0,0 +1,275 @@
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module to adjust weights spatially based on missing data in input cubes."""
+
+import iris
+import numpy as np
+
+from scipy.ndimage.morphology import distance_transform_edt
+
+from improver.utilities.rescale import rescale
+from improver.utilities.cube_checker import check_cube_coordinates
+
+
+class SpatialWeightsForMissingData(object):
+    """
+    Plugin for adjusting weights spatially based on missing data in the input
+    cube that will be collapsed using these weights.
+
+    The plugin does the following steps:
+
+        1. Creates an initial spatial weights basked on the mask in the
+           input cube giving zero weight to where there is masked data and
+           a weight of 1 where there is valid data.
+        2. Make these weights fuzzy by smoothing the boundary between where
+           there is valid data and no valid data. This keeps areas of zero
+           weight, but reduces nearby grid points with a weight of 1 depending
+           on how far they are from a grid point with a weight of zero. The
+           range of the effect is controlled by the supplied fuzzy_length
+        3. Multiplies the fuzzy spatial weights by the one_dimensional weights
+           supplied to the plugin.
+        4. Normalises the weights along the coordinatethat will be collapsed
+           when blending is carried out using these weights.
+    """
+
+    def __init__(self, fuzzy_length=10):
```

Comment:
Do you think we should do this in grid squares or kilometres?  The neighbourhood processing code does this in kilometres, which is more flexible to input file resolution, so I'd prefer that.

Reply:
Ah, I hadn't realised this.  In which case this is fine for the moment (as I think we do want this flexibility) - but I wonder whether it's worth raising follow on work to add that functionality to the neighbourhood processing code, as it seems like something we'd rather be able to cover there rather than having different conventions in different places.